File: app/src/main/java/com/coinninja/coinkeeper/cn/account/AccountManager.java
package com.coinninja.coinkeeper.cn.account;

import com.coinninja.coinkeeper.cn.wallet.HDWallet;
import com.coinninja.coinkeeper.model.db.Address;
import com.coinninja.coinkeeper.model.dto.AddressDTO;
import com.coinninja.coinkeeper.model.helpers.AddressHelper;
import com.coinninja.coinkeeper.model.helpers.WalletHelper;

import java.util.HashMap;
import java.util.List;

import javax.inject.Inject;

public class AccountManager {

    private final AddressCache addressCache;
    private final AddressHelper addressHelper;
    private final HDWallet hdWallet;
    private WalletHelper walletHelper;

    @Inject
    AccountManager(HDWallet hdWallet, WalletHelper walletHelper, AddressCache addressCache, AddressHelper addressHelper) {
        this.hdWallet = hdWallet;
        this.walletHelper = walletHelper;
        this.addressCache = addressCache;
        this.addressHelper = addressHelper;
    }

    public String getNextReceiveAddress() {
        List<Address> addresses = addressHelper.getUnusedAddressesFor(HDWallet.EXTERNAL);
        return addresses.size() > 0 ? addresses.get(0).getAddress() : hdWallet.getExternalAddress(0);
    }

    public int getNextReceiveIndex() {
        List<Address> addresses = addressHelper.getUnusedAddressesFor(HDWallet.EXTERNAL);
        return addresses.size() > 0 ? addresses.get(0).getIndex() : 0;
    }

    public int getNextChangeIndex() {
        List<Address> addresses = addressHelper.getUnusedAddressesFor(HDWallet.INTERNAL);
        return addresses.size() > 0 ? addresses.get(0).getIndex() : 0;
    }

    //TODO can find the largest consumed from DB and use that as a reference point
    @Deprecated
    public void reportLargestReceiveIndexConsumed(int index) {
        if (walletHelper.getCurrentExternalIndex() <= index) {
            walletHelper.setExternalIndex(index + 1);
        }
    }

    //TODO can find the largest consumed from DB and use that as a reference point
    @Deprecated
    public void reportLargestChangeIndexConsumed(int index) {
        if (walletHelper.getCurrentInternalIndex() <= index) {
            walletHelper.setInternalIndex(index + 1);
        }
    }

    public void cacheAddresses() {
        addressCache.cacheAddressesFor(HDWallet.EXTERNAL);
        addressCache.cacheAddressesFor(HDWallet.INTERNAL);
    }

    public HashMap<String, AddressDTO> unusedAddressesToPubKey(int chainIndex, int blockSize) {
        List<Address> unusedAddresses = addressHelper.getUnusedAddressesFor(chainIndex);
        HashMap<String, AddressDTO> addressToDTO = new HashMap<String, AddressDTO>();

        blockSize = blockSize <= unusedAddresses.size() ? blockSize : unusedAddresses.size();

        for (int i = 0; i < blockSize; i++) {
            Address address = unusedAddresses.get(i);
            String uncompressedPubKey = addressCache.getUncompressedPublicKey(address.getDerivationPath());
            addressToDTO.put(address.getAddress(), new AddressDTO(address, uncompressedPubKey));
        }

        return addressToDTO;
    }

    public int getLargestReportedChangeAddress() {
        return addressHelper.getLargestDerivationIndexReportedFor(HDWallet.INTERNAL);
    }

    public int getLargestReportedReceiveAddress() {
        return addressHelper.getLargestDerivationIndexReportedFor(HDWallet.EXTERNAL);
    }
}

File: app/src/main/java/com/coinninja/coinkeeper/cn/account/AccountManager.kt
package com.coinninja.coinkeeper.cn.account

import app.coinninja.cn.libbitcoin.model.DerivationPath
import app.dropbit.annotations.Mockable
import com.coinninja.coinkeeper.cn.wallet.HDWalletWrapper
import com.coinninja.coinkeeper.model.db.Wallet
import com.coinninja.coinkeeper.model.dto.AddressDTO
import com.coinninja.coinkeeper.model.helpers.AddressHelper
import com.coinninja.coinkeeper.model.helpers.WalletHelper
import javax.inject.Inject

@Mockable
class AccountManager @Inject internal constructor(
        internal val hdWallet: HDWalletWrapper,
        internal val walletHelper: WalletHelper,
        internal val addressCache: AddressCache,
        internal val addressHelper: AddressHelper
) {

    val nextReceiveAddress: String
        get() {
            val addresses = addressHelper.getUnusedAddressesFor(walletHelper.primaryWallet, HDWalletWrapper.EXTERNAL)
            return if (addresses.isNotEmpty()) addresses[0].address else hdWallet.getAddressForPath(derivationPathForExternalIndex(0)).address
        }

    val nextReceiveIndex: Int
        get() {
            val addresses = addressHelper.getUnusedAddressesFor(walletHelper.primaryWallet, HDWalletWrapper.EXTERNAL)
            return if (addresses.isNotEmpty()) addresses[0].index else 0
        }

    val nextChangeIndex: Int
        get() {
            val addresses = addressHelper.getUnusedAddressesFor(walletHelper.primaryWallet, HDWalletWrapper.INTERNAL)
            return if (addresses.isNotEmpty()) addresses[0].index else 0
        }

    fun largestReportedChangeAddress(wallet: Wallet): Int = addressHelper.getLargestDerivationIndexReportedFor(wallet, HDWalletWrapper.INTERNAL)

    fun largestReportedReceiveAddress(wallet: Wallet): Int = addressHelper.getLargestDerivationIndexReportedFor(wallet, HDWalletWrapper.EXTERNAL)

    private fun derivationPathForExternalIndex(index: Int): DerivationPath {
        val wallet = walletHelper.primaryWallet
        return DerivationPath(wallet.purpose, wallet.coinType, wallet.accountIndex, HDWalletWrapper.EXTERNAL, index)
    }

    fun reportLargestReceiveIndexConsumed(wallet: Wallet, index: Int) {
        if (wallet.externalIndex <= index) {
            wallet.externalIndex = index
            wallet.update()
        }
    }

    fun reportLargestChangeIndexConsumed(wallet: Wallet, index: Int) {
        if (wallet.internalIndex <= index) {
            wallet.internalIndex = index
            wallet.update()
        }
    }

    fun cacheAddresses(wallet: Wallet) {
        addressCache.cacheAddressesFor(wallet, HDWalletWrapper.EXTERNAL)
        addressCache.cacheAddressesFor(wallet, HDWalletWrapper.INTERNAL)
    }

    fun unusedAddressesToPubKey(chainIndex: Int, blockSize: Int): HashMap<String, AddressDTO> {
        val unusedAddresses = addressHelper.getUnusedAddressesFor(walletHelper.primaryWallet, chainIndex)
        val addressToDTO = HashMap<String, AddressDTO>()
        val size = if (blockSize <= unusedAddresses.size) blockSize else unusedAddresses.size

        for (i in 0 until size) {
            val address = unusedAddresses[i]
            addressToDTO[address.address] = AddressDTO(address, addressCache.getUncompressedPublicKey(address.derivationPath))
        }

        return addressToDTO
    }
}
